core/WFCORE-5343_Support_YAML_files_for_configuration_customization.adoc

```diff
@@ -0,0 +1,164 @@
+= Support for YAML files to customize WildFly configuration
+:author:            Emmanuel Hugonnet
+:email:             ehugonne@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+The aim of this proposal is to be able to inject external configuration in a YAML format to WildFly in order to customize an existing configuration.
```

Comment:
We see overlap between this feature and https://issues.redhat.com/browse/WFCORE-5324.

Can you elaborate in proposal what does it add compared to https://issues.redhat.com/browse/WFCORE-5324. Is it just different format? I assume underneath both cli and yaml ways end in DMR so behaviour, starting times and limitations would be same.

Can you elaborate on that as well?

Reply:
It is different in that the operations from the YAML files are executed on boot in addition to the one from the XML file. So this means that we should have better performance, we don't require a reload of the server and we don't change the resulting XML configuration since we are in "read-only" mode on that part.

Reply:
Also this isn't a script: no if/else etc. 
Also you can reuse the same command line in a loop without having to have the logic in your script to manage existing resources.
Last but not least, this is for all versions of WildFly: bootable and bare metal.

Reply:
Can I achieve `better performance` and `no reload requirement` and `idempotency` with CLI as well? Or is it consequence of limited scope of yaml files? Still I assume `CLI way` must be superset of `yaml way`, because in the end yaml must use same capabilities as CLI. Am I missing something?

So I have cli way which is applicable everywhere; bootable jar, bare metal, openshift, domain. And I can express anything in CLI. 

I will have yaml file with much more limited usage (just bootable jar, bare metal, just subset of CLI capabilities). 

Why should I use yaml file? When I will be writing yaml files I will need to try cli command before transform it into yaml snippet anyway. Isn't it better to directly compose cli script?

I agree yaml file is easy to read. But something similar can be achieved with formatting of CLI commands. ( I know it is not yaml :) )
```
/authentication=classic/login-module=Remoting:add( \
    code=Remoting, \
    flag=optional, \
    module-options=[ \
        password-stacking=useFirstPass \
    ] \
)
```

Reply:
It would work on OpenShift ;-) Just domain is not supported.
You won't test using CLI script, in the same way you don't test using CLI to write or configure a feature in Galleon.
CLI script requires a reload: so boot-time is longer. Like I wrote the cost is almost just the cost of parsing the YAML.
In your case you are missing the test to check if the resource already exists. Also you would have to create all the tree, while here we would create them for you.

Reply:
@mchoma, yaml syntax, in the context of bootable JAR, is to help transitioning from Thorntail to WildFly Bootable JAR. The object model it exposed can be seen as simpler than CLI syntax. CLI implies a start in admin and a reload. We apply CLI commands (that are more than DMR) on top of a running server. Yaml will be more efficient. Having both is powerful. The order of application is: 1) yaml files 2) CLI script. With CLI you can call some management operations not exposed by YAML. I suspect that yaml will be much more used than CLI with bootable JAR. 
I don't know of its usage outside of bootable JAR. If yaml is usable outside of Bootable JAR, we could have an issue with Openshift s2i images. The YAML being applied first, the server would be not yet fully configured by the CLI script generated by our launch scripts (that currently relies on the same mechanism as the one used in WFCORE-5324).